### PR TITLE
Vulkan: handles uniform ubo command i.e., add UBO

### LINKED ISSUE #147: Add UBO which handles uniform ubo command of VkRunner

### PR DESCRIPTION
**Pull request by jaebaek**
Fixes #147

### CODE REVIEW

File: src/vulkan/buffer_descriptor.h [hunk 1]
```diff
@@ -29,12 +29,15 @@
 namespace amber {
 namespace vulkan {
 
-class StorageBufferDescriptor : public Descriptor {
+class BufferDescriptor : public Descriptor {
```
[the remark this answers is not in the document]
Comment (jaebaek) [reply src/vulkan/buffer_descriptor.h]: done

File: src/vulkan/buffer_descriptor.h [hunk 2]
```diff
@@ -29,12 +29,15 @@
...
-  StorageBufferDescriptor(VkDevice device, uint32_t desc_set, uint32_t binding);
-  ~StorageBufferDescriptor();
+  BufferDescriptor(DescriptorType type,
+                   VkDevice device,
+                   uint32_t desc_set,
+                   uint32_t binding);
+  ~BufferDescriptor();
 
-  void FillBufferWithData(const SSBOData& data);
+  void FillBufferWithData(const BufferData& data);
```
[the remark this answers is not in the document]
Comment (jaebaek) [reply src/vulkan/buffer_descriptor.h]: done

File: src/vulkan/descriptor.h [hunk 3]
```diff
@@ -114,7 +114,7 @@ class Descriptor {
       VkCommandBuffer command,
       const VkPhysicalDeviceMemoryProperties& properties) = 0;
 
-  // Record a command for copying data in |ssbo_data_queue_| to the
+  // Record a command for copying data in |buffer_data_queue_| to the
```
[the remark this answers is not in the document]
Comment (jaebaek) [reply src/vulkan/descriptor.h]: By "to the resource", I meant "to device". I changed it as "to the resource in device".

File: src/vulkan/descriptor.h [hunk 4]
```diff
@@ -140,11 +140,11 @@ class Descriptor {
 
   VkDevice GetDevice() const { return device_; }
 
-  const std::vector<SSBOData>& GetSSBODataQueue() const {
-    return ssbo_data_queue_;
+  const std::vector<BufferData>& GetSSBODataQueue() const {
```
[the remark this answers is not in the document]
Comment (jaebaek) [reply src/vulkan/descriptor.h]: done

File: src/vulkan/descriptor.h [hunk 5]
```diff
@@ -140,11 +140,11 @@ class Descriptor {
...
+    return buffer_data_queue_;
   }
 
-  void ClearSSBODataQueue() { ssbo_data_queue_.clear(); }
+  void ClearSSBODataQueue() { buffer_data_queue_.clear(); }
```
[the remark this answers is not in the document]
Comment (jaebaek) [reply src/vulkan/descriptor.h]: Done. Note that it creates separate objects for different SSBO and UBO.
For example, if there are SSBO with desc set 0, binding 0, SSBO with 0, 2, and UBO with 2, 1, it creates three unique_ptr<BufferDescriptor> objects. It just has a wrong name.

File: src/vulkan/descriptor.h [hunk 6]
```diff
@@ -164,7 +164,7 @@ class Descriptor {
 
   DescriptorType type_ = DescriptorType::kSampledImage;
   VkDevice device_ = VK_NULL_HANDLE;
-  std::vector<SSBOData> ssbo_data_queue_;
+  std::vector<BufferData> buffer_data_queue_;
```
[the remark this answers is not in the document]
Comment (jaebaek) [reply src/vulkan/descriptor.h]: done

File: src/vulkan/engine_vulkan.cc [hunk 7]
```diff
@@ -355,8 +355,8 @@ Result EngineVulkan::GetDescriptorInfo(const uint32_t descriptor_set,
 }
 
 Result EngineVulkan::DoBuffer(const BufferCommand* command) {
-  if (!command->IsSSBO())
-    return Result("Vulkan::DoBuffer non-SSBO descriptor not implemented");
+  if (command->IsPushConstant())
```
[the remark this answers is not in the document]
Comment (jaebaek) [reply src/vulkan/engine_vulkan.cc]: done

File: src/vulkan/pipeline.cc [hunk 8]
```diff
@@ -238,8 +238,8 @@ Result Pipeline::UpdateDescriptorSetsIfNeeded() {
 }
 
 Result Pipeline::AddDescriptor(const BufferCommand* buffer_command) {
-  if (!buffer_command->IsSSBO())
-    return Result("Vulkan::AddDescriptor non-SSBO not implemented");
+  if (buffer_command->IsPushConstant())
```
[the remark this answers is not in the document]
Comment (jaebaek) [reply src/vulkan/pipeline.cc]: done

File: src/vulkan/pipeline.cc [hunk 9]
```diff
@@ -251,17 +251,37 @@ Result Pipeline::AddDescriptor(const BufferCommand* buffer_command) {
...
+          "Vulkan::AddDescriptor BufferCommand for SSBO uses wrong descriptor "
+          "set and binding");
+
+    desc->AddToBufferDataQueue(
+        buffer_command->GetDatumType().GetType(), buffer_command->GetOffset(),
+        buffer_command->GetSize(), buffer_command->GetValues());
+  }
+
+  if (buffer_command->IsUniform()) {
+    if (!desc->IsUniformBuffer())
```
[the remark this answers is not in the document]
Comment (jaebaek) [reply src/vulkan/pipeline.cc]: done

File: src/vulkan/pipeline.cc [hunk 10]
```diff
@@ -251,17 +251,39 @@ Result Pipeline::AddDescriptor(const BufferCommand* buffer_command) {
...
-      buffer_command->GetDatumType().GetType(), buffer_command->GetOffset(),
-      buffer_command->GetSize(), buffer_command->GetValues());
+  if (buffer_command->IsSSBO()) {
+    if (!desc->IsStorageBuffer()) {
+      return Result(
+          "Vulkan::AddDescriptor BufferCommand for SSBO uses wrong descriptor "
+          "set and binding");
+    }
+
+    desc->AddToBufferDataQueue(
```
Comment (dj2) [inline src/vulkan/pipeline.cc]: This update is the same for both if's so pull it out and do it after the second if()

File: src/vulkan/pipeline.cc [hunk 11]
```diff
@@ -251,17 +251,39 @@ Result Pipeline::AddDescriptor(const BufferCommand* buffer_command) {
...
+    descriptors_.push_back(std::move(buffer_desc));
 
     desc = descriptors_.back().get();
   }
 
-  desc->AddToSSBODataQueue(
-      buffer_command->GetDatumType().GetType(), buffer_command->GetOffset(),
-      buffer_command->GetSize(), buffer_command->GetValues());
+  if (buffer_command->IsSSBO()) {
+    if (!desc->IsStorageBuffer()) {
```
Comment (dj2) [inline src/vulkan/pipeline.cc]: Once the update is pulled out these two if's can be combined (same with below)

Comment (jaebaek) [reply src/vulkan/pipeline.cc]: done

File: src/vulkan/pipeline.cc (hunk 10 above)
Comment (jaebaek) [reply src/vulkan/pipeline.cc]: done

**Comment by jaebaek** [comment]
It should be reviewed by @dneto0 if there are any concerns later, but now I will merge it

**Pull request merged**